rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,480 @@
+:rewardConfigIndex = read(rewardConfigCount) - 1;
+repeat :rewardConfig = read(rewardConfigs[rewardConfigIndex--]);
+    :reward_allocation = Assets::balance_of(reward_allocation_acc,
+            rewardConfig.asset);
+    :epochReward = reward_allocation * rewardConfig.reward_rate;
+    :stakerReward = epochReward;
+    :perAssetRewardAllocationIndex = read(perAssetRewardAllocationCount) - 1;
+    repeat :lpRewardAllocation = read(perAssetRewardAllocation[perAssetRewardAllocationIndex--]);
+        :lpEpochReward = EpochReward * lpRewardAllocation;
+        :staking_rewards::transfer_reward(
+            lpRewardAllocation.asset,
+            rewardConfig.asset,
+            reward_allocation_acc,
+            lpEpochReward, false);
+        :stakerReward -= lpEpochReward;
+    backward:perAssetRewardAllocationIndex;
+    repeat while (perAssetRewardAllocationIndex != -1)
+    :staking_rewards::transfer_reward(
+        rewardConfig.asset,
+        rewardConfig.asset,
+        reward_allocation_acc,
+        stakerReward, false);
+    note left: remainder after the per\nasset allocation is\ntransferred to same\nasset stakers reward
+backward:rewardConfigIndex;
+repeat while (rewardConfigIndex != -1)
+stop
+----
+
+This algorithm runs in `O(staked_asset_type_count * rewarded_asset_type_count)`.
+
+== Implementation
+
+=== Pallet Pablo: LP Fee + Staking Changes
+
+- [ ] Implement <<_feeconfig>> on pallet-pablo across all 3 types of pools.
+- [ ] Implement <<_pblo_staker_trading_fee_distribution>>.
+
+=== Pallet Staking Rewards: PICA/PBLO Staking Related Changes
+
+- [ ] Implement <<_rewardaccumulationhook>>.
+
+[appendix]
+== Fee Distribution Q&A
+
+Based on the current setup following questions arise when deciding on the distribution of these fees to relevant liquidity providers, owners and stakers.
+
+. A Protocol Fee for all pools in Pablo (or even protocol pallets other than Pablo)?
++
+Does it make sense to define a protocol fee percentage on top of the pool owner fees of the pools so that the protocol fee can be used as the pot out of which the stakers are rewarded? Initially the Protocol Fee = Pool Owner Fee as the pools are owned by Composable. Assumption here is that the stakers would indeed still get a reward out of third party created pool fees.
++
+*Comment:* While having a protocol funding mechanism is valuable, initially the protocol fees should zero or minimal.
+. How does the system reward PICA stakers? Wouldn't the Pablo protocol needs some parameter to define how much of its swap fee or protocol fee as referred to above would go to PICA holders? Or do we assume that PICA stakers do not get a reward out of the Pablo pool fees?
+.. If Pablo does reward PICA stakers, the system might need a common interface that directs those funds out of Pablo.
+.. If Pablo does reward PICA stakers, the system might need to have a treasury parameter that defines the percentage that goes out to PICA holders that can be adjusted overtime.
```

Comment:
each pool can define ratio of lp to pica to pablo. and split revenue here. at same time it defines ration, it may define 3 protocol ids and store array in pablo. when palo rewards, it loops over pools.